README.md
<big><h1 align="center">loopback-nodejs-client</h1></big>

<p align="center">
  <a href="https://npmjs.org/package/loopback-nodejs-client">
    <img src="https://img.shields.io/npm/v/loopback-nodejs-client.svg?style=flat-square"
         alt="NPM Version">
  </a>

  <a href="https://coveralls.io/r/sguilly/loopback-nodejs-client">
    <img src="https://img.shields.io/coveralls/sguilly/loopback-nodejs-client.svg?style=flat-square"
         alt="Coverage Status">
  </a>

  <a href="https://travis-ci.org/sguilly/loopback-nodejs-client">
    <img src="https://img.shields.io/travis/sguilly/loopback-nodejs-client.svg?style=flat-square"
         alt="Build Status">
  </a>

  <a href="https://npmjs.org/package/loopback-nodejs-client">
    <img src="http://img.shields.io/npm/dm/loopback-nodejs-client.svg?style=flat-square"
         alt="Downloads">
  </a>

  <a href="https://david-dm.org/sguilly/loopback-nodejs-client.svg">
    <img src="https://david-dm.org/sguilly/loopback-nodejs-client.svg?style=flat-square"
         alt="Dependency Status">
  </a>

  <a href="https://github.com/sguilly/loopback-nodejs-client/blob/master/LICENSE">
    <img src="https://img.shields.io/npm/l/loopback-nodejs-client.svg?style=flat-square"
         alt="License">
  </a>
</p>

<p align="center"><big>

</big></p>

> NodeJS client library with loopback SDK syntax


## Install

```sh
$ npm install --save loopback-nodejs-client
```


## Usage

### Without authentification

```js
var LoopbackClient = require('loopback-nodejs-client')

var loopbackClient = new LoopbackClient('http://localhost:3000/api');

var Customers = loopbackClient.getModel('customers')

Customers.create({name: 'ido4pro', city: 'plougonvelin'})
  .then(function (customer) {
    console.log(customer)

    Customers.find({filter: {where: {name: 'ido4pro'}}})
      .then(function (customers) {
        console.log(customers)

      })
      .catch(function (err) {
        console.log(err)
      })

    Customers.count({where: {name: 'ido4pro'}})
      .then(function (customers) {
        console.log(customers)

      })
      .catch(function (err) {
        console.log(err)
      })

  })
  .catch(function (err) {
    console.log(err)
  })
```

### With authentification

```js
var loopbackClientWithAuth = new LoopbackClient('http://localhost:3000/api','xxx EMAIL USER xxx', 'xxx PASSWORD xxx')

loopbackClientWithAuth.createToken()
  .then(function()
  {
      
       var Customers = loopbackClientWithAuth.getModel('customers')
       
       Customers.count({where: {name: 'ido4pro'}})
            .then(function (customers) {
              console.log(customers)
      
            })
            .catch(function (err) {
              console.log(err)
            })
      
  })


```

## Methods
  
  create(data) --> data = { id: 'xxx', filter: {...} } (filter is optional)

  count(where) --> where = { where : {name: 'xxx'} }, where = { where: {and: [{name: 'xxx'},{city: 'xxx'}] 

  updateAll(where,data) 

  updateById(id,where)

  find(filter)
   
  findById(data)

  findOne(query)

  deleteById(id)
  
  
  Check the test/index.js for more usage
  

## License

MIT © [Stéphane GUILLY](http://github.com/sguilly)

[npm-url]: https://npmjs.org/package/loopback-nodejs-client
[npm-image]: https://img.shields.io/npm/v/loopback-nodejs-client.svg?style=flat-square

[travis-url]: https://travis-ci.org/sguilly/loopback-nodejs-client
[travis-image]: https://img.shields.io/travis/sguilly/loopback-nodejs-client.svg?style=flat-square

[coveralls-url]: https://coveralls.io/r/sguilly/loopback-nodejs-client
[coveralls-image]: https://img.shields.io/coveralls/sguilly/loopback-nodejs-client.svg?style=flat-square

[depstat-url]: https://david-dm.org/sguilly/loopback-nodejs-client
[depstat-image]: https://david-dm.org/sguilly/loopback-nodejs-client.svg?style=flat-square

[download-badge]: http://img.shields.io/npm/dm/loopback-nodejs-client.svg?style=flat-square


CHANGELOG.md
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
# Changelog

- [v1.0.0](#v100)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

### v1.0.0

 * rise-module initial commit.


test/index.js
var test = require('tape')
var LoopbackClient = require('../dist')

// to test the es6 version
//import test from "tape"
//import LoopbackClient from "../src"


const user = 'stephane.guilly@gmail.com'
const password = 'xxxx'

var nock = require('nock');

//nock.recorder.rec();

var loopbackClient;

var idCustomer;


test("token", (t) => {
   t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .post('/api/users/login', {"email":user,"password":password})
    .query({"include":"user"})
    .reply(200,
      {"id":"aieLGPjVF3hAJmQivH74FvqEbuzpteqfJ30NC1mtWvgMA584GD35fQ6BLoMNh6Ds",
        "ttl":1209600,
        "created":"2017-01-03T12:51:41.687Z",
        "userId":"5752f3cde8e5812c2592eebb",
      }, [ 'Vary',
        'Origin, Accept-Encoding',
        'Access-Control-Allow-Credentials',
        'true',
        'X-XSS-Protection',
        '1; mode=block',
        'X-Frame-Options',
        'DENY',
        'X-Download-Options',
        'noopen',
        'X-Content-Type-Options',
        'nosniff',
        'Content-Type',
        'application/json; charset=utf-8',
        'Content-Length',
        '915',
        'ETag',
        'W/"393-wbs82x3xuxWwFOsv4/Atdw"',
        'Date',
        'Tue, 03 Jan 2017 12:51:41 GMT',
        'Connection',
        'close' ]);

  loopbackClient = new LoopbackClient('http://localhost:42001/api',user,password);

   console.log('loopbackClient',loopbackClient)

  loopbackClient.createToken().then((token)=>{
    console.log('token=',token)

    t.equal(true,token.length > 0)
  })

})

test("create", (t) => {
  t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .post('/api/customers', {"name":"ido4pro","city":"plougonvelin"})
    .reply(200, {"name":"ido4pro","id":"586b9fe866a6ea6347c290ae","city":"plougonvelin"}, [ 'Vary',
      'Origin, Accept-Encoding',
      'Access-Control-Allow-Credentials',
      'true',
      'X-XSS-Protection',
      '1; mode=block',
      'X-Frame-Options',
      'DENY',
      'X-Download-Options',
      'noopen',
      'X-Content-Type-Options',
      'nosniff',
      'Content-Type',
      'application/json; charset=utf-8',
      'Content-Length',
      '72',
      'ETag',
      'W/"48-c0ddKdkGIDvb+MDPRWRwmg"',
      'Date',
      'Tue, 03 Jan 2017 12:58:16 GMT',
      'Connection',
      'close' ]);

  var Customers = loopbackClient.getModel('customers')

  Customers.create({name: 'ido4pro', city: 'plougonvelin'})
    .then(function (customer) {


      t.equal('ido4pro',customer.name)

      idCustomer = customer.id

    })
    .catch(function (err) {
      console.log(err)
    })

})

test("find", (t) => {
  t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .get('/api/customers')
    .query(function(actualQueryObject){
      // do some compare with the actual Query Object
      // return true for matched
      // return false for not matched

      console.log('actualQueryObject',actualQueryObject)
      return actualQueryObject.filter.where.name == 'ido4pro' ? true : false  ;
    })
    .reply(200, [{"name":"ido4pro","id":idCustomer,"city":"plougonvelin"}], [ 'Vary',
      'Origin, Accept-Encoding',
      'Access-Control-Allow-Credentials',
      'true',
      'X-XSS-Protection',
      '1; mode=block',
      'X-Frame-Options',
      'DENY',
      'X-Download-Options',
      'noopen',
      'X-Content-Type-Options',
      'nosniff',
      'Content-Type',
      'application/json; charset=utf-8',
      'Content-Length',
      '585',
      'ETag',
      'W/"249-BIjYMpLoc7unccLIspFwog"',
      'Date',
      'Tue, 03 Jan 2017 13:09:54 GMT',
      'Connection',
      'close' ]);




  var Customers = loopbackClient.getModel('customers')


      Customers.find({filter: {where: {name: 'ido4pro'}}})
        .then(function (customers) {
          t.equal(true,customers.length > 0 )

        })
        .catch(function (err) {
          console.log(err)
        })

})

test("findOne", (t) => {
  t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .get('/api/customers/findOne')
    .query(function(actualQueryObject){
      // do some compare with the actual Query Object
      // return true for matched
      // return false for not matched

      console.log('actualQueryObject',actualQueryObject)
      return actualQueryObject.filter.where.name == 'ido4pro' ? true : false  ;
    })
    .reply(200, {"name":"ido4pro","id":idCustomer,"city":"plougonvelin"}, [ 'Vary',
      'Origin, Accept-Encoding',
      'Access-Control-Allow-Credentials',
      'true',
      'X-XSS-Protection',
      '1; mode=block',
      'X-Frame-Options',
      'DENY',
      'X-Download-Options',
      'noopen',
      'X-Content-Type-Options',
      'nosniff',
      'Content-Type',
      'application/json; charset=utf-8',
      'Content-Length',
      '585',
      'ETag',
      'W/"249-BIjYMpLoc7unccLIspFwog"',
      'Date',
      'Tue, 03 Jan 2017 13:09:54 GMT',
      'Connection',
      'close' ]);




  var Customers = loopbackClient.getModel('customers')


  Customers.findOne({filter: {where: {name: 'ido4pro'}}})
  //Customers.findOne({where: {name: 'ido4pro'}})
    .then(function (customer) {
      t.equal(true,customer.name == 'ido4pro' )

    })
    .catch(function (err) {
      console.log(err)
    })

})

test("findById", (t) => {
  t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .get('/api/customers/'+idCustomer)
    .reply(200, {"name":"ido4pro","id":idCustomer,"city":"plougonvelin"}, [ 'Vary',
      'Origin, Accept-Encoding',
      'Access-Control-Allow-Credentials',
      'true',
      'X-XSS-Protection',
      '1; mode=block',
      'X-Frame-Options',
      'DENY',
      'X-Download-Options',
      'noopen',
      'X-Content-Type-Options',
      'nosniff',
      'Content-Type',
      'application/json; charset=utf-8',
      'Content-Length',
      '72',
      'ETag',
      'W/"48-c0ddKdkGIDvb+MDPRWRwmg"',
      'Date',
      'Tue, 03 Jan 2017 13:24:00 GMT',
      'Connection',
      'close' ]);

  var Customers = loopbackClient.getModel('customers')


  Customers.findById({ id: idCustomer})
    .then(function (customer) {
      t.equal(true,customer.name == 'ido4pro')

    })
    .catch(function (err) {
      console.log(err)
    })

})


test("updateById", (t) => {
  t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .put('/api/customers/'+idCustomer, {"city":"brest"})
    .reply(200, {"name":"ido4pro","id":idCustomer,"city":"brest"}, [ 'Vary',
      'Origin, Accept-Encoding',
      'Access-Control-Allow-Credentials',
      'true',
      'X-XSS-Protection',
      '1; mode=block',
      'X-Frame-Options',
      'DENY',
      'X-Download-Options',
      'noopen',
      'X-Content-Type-Options',
      'nosniff',
      'Content-Type',
      'application/json; charset=utf-8',
      'Content-Length',
      '65',
      'ETag',
      'W/"41-c4/OB+yDE5yGjChOibGgTw"',
      'Date',
      'Wed, 04 Jan 2017 09:13:39 GMT',
      'Connection',
      'close' ]);



  var Customers = loopbackClient.getModel('customers')


  Customers.updateById(idCustomer,{city: 'brest'})
    .then(function (customer) {
      t.equal(true,customer.city == 'brest')

    })
    .catch(function (err) {
      console.log(err)
    })

})

test("updateAll", (t) => {
  t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .post('/api/customers/update', {"city":"nantes"})
    .query(function(actualQueryObject){
      // do some compare with the actual Query Object
      // return true for matched
      // return false for not matched

      console.log('actualQueryObject',actualQueryObject)
      return actualQueryObject.where.name == 'ido4pro' ? true : false  ;
    })
    .reply(200, {"count":1}, [ 'Vary',
      'Origin, Accept-Encoding',
      'Access-Control-Allow-Credentials',
      'true',
      'X-XSS-Protection',
      '1; mode=block',
      'X-Frame-Options',
      'DENY',
      'X-Download-Options',
      'noopen',
      'X-Content-Type-Options',
      'nosniff',
      'Content-Type',
      'application/json; charset=utf-8',
      'Content-Length',
      '12',
      'ETag',
      'W/"c-d07cdU1Uym2u//VIRStBDQ"',
      'Date',
      'Wed, 04 Jan 2017 16:19:16 GMT',
      'Connection',
      'close' ]);



  var Customers = loopbackClient.getModel('customers')


  Customers.updateAll({where:{name:'ido4pro'}},{city: 'nantes'})
    .then(function (value) {
      t.equal(true,value.count > 0)

    })
    .catch(function (err) {
      console.log(err)
    })

})

test("delete", (t) => {
  t.plan(1)

  nock('http://localhost:42001', {"encodedQueryParams":true})
    .delete('/api/customers/'+idCustomer)
    .reply(200, {"count":1}, [ 'Vary',
      'Origin, Accept-Encoding',
      'Access-Control-Allow-Credentials',
      'true',
      'X-XSS-Protection',
      '1; mode=block',
      'X-Frame-Options',
      'DENY',
      'X-Download-Options',
      'noopen',
      'X-Content-Type-Options',
      'nosniff',
      'Content-Type',
      'application/json; charset=utf-8',
      'Content-Length',
      '11',
      'ETag',
      'W/"b-CxuxFwe2mm4IC62zgWSn6Q"',
      'Date',
      'Tue, 03 Jan 2017 13:31:19 GMT',
      'Connection',
      'close' ]);


  var Customers = loopbackClient.getModel('customers')


  Customers.deleteById(idCustomer)
    .then(function (value) {
      t.equal(true,value.count == 1)

    })
    .catch(function (err) {
      console.log(err)
    })

})


gulpfile.js
'use strict';
var gulp = require('gulp');
var eslint = require('gulp-eslint');
var excludeGitignore = require('gulp-exclude-gitignore');
var mocha = require('gulp-mocha');
var istanbul = require('gulp-istanbul');
var nsp = require('gulp-nsp');
var plumber = require('gulp-plumber');
var babel = require('gulp-babel');
var del = require('del');
var isparta = require('isparta');

// Initialize the babel transpiler so ES2015 files gets compiled
// when they're loaded
require('babel-core/register');

gulp.task('static', function () {
  return gulp.src('**/*.js')
    .pipe(excludeGitignore())
    .pipe(eslint())
    .pipe(eslint.format())
    .pipe(eslint.failAfterError());
});

gulp.task('nsp', function (cb) {
  nsp('package.json', cb);
});

gulp.task('pre-test', function () {
  return gulp.src('lib/**/*.js')
    .pipe(istanbul({
      includeUntested: true
,      instrumenter: isparta.Instrumenter
    }))
    .pipe(istanbul.hookRequire());
});

gulp.task('test', ['pre-test'], function (cb) {
  var mochaErr;

  gulp.src('test/**/*.js')
    .pipe(plumber())
    .pipe(mocha({reporter: 'spec'}))
    .on('error', function (err) {
      mochaErr = err;
    })
    .pipe(istanbul.writeReports())
    .on('end', function () {
      cb(mochaErr);
    });
});

gulp.task('babel', ['clean'], function () {
  return gulp.src('lib/**/*.js')
    .pipe(babel())
    .pipe(gulp.dest('dist'));
});

gulp.task('clean', function () {
  return del('dist');
});

gulp.task('prepublish', ['nsp', 'babel']);
gulp.task('default', ['static', 'test']);


lib/LoopbackClient.js
/**
 * Created by sguilly on 17/11/16.
 */
/**
 * Created by sguilly on 03/11/16.
 */
"use strict";

var rest = require('restler');
var LoopbackModel = require('./LoopbackModel.js');

class LoopbackClient {
  constructor(baseUrl,user,password) {

    this.user = user
    this.password = password
    this.token = null
    this.baseUrl = baseUrl
  }

  getBaseUrl()
  {
    return this.baseUrl
  }

  getToken()
  {
    console.log('1=',this.token)

    return this.token
  }

  createToken()
  {
    var promise = new Promise((resolve, reject) => {

      if(this.token)
      {
        console.log('token from cache')
        resolve(this.token);
      }
      else {

        console.log('create token')

        var data = {
          email: this.user,
          password: this.password
        }

        rest.postJson(this.baseUrl+'/users/login?include=user', data

        ).on('complete', (result) =>{

          //console.log('result=',result)
          if (result instanceof Error) {
            reject(result.message);

          } else {
            this.token = result.id
            resolve(this.token);
          }
        });

      }

    })

    return promise
  }

  getModel(name)
  {
    return new LoopbackModel(name,this)
  }
}


module.exports = LoopbackClient


src/LoopbackModel.js
/**
 * Created by sguilly on 17/11/16.
 */
/**
 * Created by sguilly on 03/11/16.
 */
"use strict";

var rest = require('restler');

class LoopbackModel {
  constructor(model,tokenClient) {

    this.baseUrl = tokenClient.getBaseUrl()

    this.authorization = {authorization: tokenClient.getToken()}

    this.model = model
  }

  get(url, query) {

    var promise = new Promise((resolve, reject) => {

      rest.get(url,
        {
          headers: this.authorization,
          query: query
        }
      ).on('complete', function (result,response) {
        if (result instanceof Error) {
          reject(result.message);

        } else {
          if(response.statusCode != 200)
          {
            reject(result)
          }
          else
          {
            resolve(result);
          }

        }
      });

    })
    return promise
  }

  post(url, data,query) {

    console.log('url', url)

    var promise = new Promise( (resolve, reject) => {

      var options = {
        headers: this.authorization
      }

      if(query)
      {
        options.query = query
      }

      rest.postJson(url, data, options

      ).on('complete', function (result,response) {

        console.log('result=',result)
        if (result instanceof Error) {
          reject(result.message);

        } else {
          if(response.statusCode != 200)
          {
            reject(result)
          }
          else
          {
            resolve(result);
          }

        }
      });

    });

    return promise

  }

  put(url, data) {

    console.log('url', url)

    var promise = new Promise( (resolve, reject) => {

      rest.putJson(url, data,
        {
          headers: this.authorization
        }
      ).on('complete', function (result,response) {
        if (result instanceof Error) {
          reject(result.message);

        } else {
          if(response.statusCode != 200)
          {
            reject(result)
          }
          else
          {
            resolve(result);
          }

        }
      });

    });

    return promise

  }

  del(url) {

    console.log('url', url)

    var promise = new Promise( (resolve, reject) => {

      rest.del(url,
        {
          headers: this.authorization
        }
      ).on('complete', function (result,response) {
        if (result instanceof Error) {
          reject(result.message);

        } else {
          if(response.statusCode != 200)
          {
            reject(result)
          }
          else
          {
            resolve(result);
          }

        }
      });

    });

    return promise

  }

  findById(data) {

    var url = `${this.baseUrl}/${this.model}/${data.id}`
    return this.get(url,  {filter: data.filter})
  }

  create(data) {
    var url = `${this.baseUrl}/${this.model}`
    return this.post(url,data)
  }

  count(where) {
    var url = `${this.baseUrl}/${this.model}/count`
    return this.get(url,where)
  }

  updateAll(query, data) {
    var url = `${this.baseUrl}/${this.model}/update`
    return this.post(url,data,query)
  }

  updateById(id,data) {
    var url = `${this.baseUrl}/${this.model}/${id}`
    return this.put(url,data)
  }

  find(filter) {
    var url = `${this.baseUrl}/${this.model}`
    return this.get(url,filter)
  }

  findOne(query) {
    var url = `${this.baseUrl}/${this.model}/findOne`
    return this.get(url,query)
  }

  deleteById(id) {
    var url = `${this.baseUrl}/${this.model}/${id}`
    return this.del(url)
  }

}


module.exports = LoopbackModel
